### PR TITLE
[DPE-5005] Update charm to latest standards

### PR DESCRIPTION
**Pull request by Gu1nness**
## Issue

* Charm isn't up to latest standards:

## Solution 
* Update libraries
* Update workflows
* Update tox.ini
* Fix a flaky test
* Use juju 3.4.X
* Library `mongos_client_interface` is not published.

### CODE REVIEW

**Review by MiaAltieri — DISMISSED** [review]
LGTM don't forget to do these things to finalise your work with Allure
- for allure, remember to add branch: https://github.com/canonical/data-platform-workflows/blob/main/.github/workflows/integration_test_charm_allure_beta.md
- remember to add to charmcraftcache-hub: https://github.com/canonical/data-platform-workflows/blob/main/.github/workflows/build_charm.md

**Comment by Gu1nness** [comment]
> for allure, remember to add branch: [canonical/data-platform-workflows@main/.github/workflows/integration_test_charm_allure_beta.md](https://github.com/canonical/data-platform-workflows/blob/main/.github/workflows/integration_test_charm_allure_beta.md?rgh-link-date=2024-08-08T09%3A08%3A50Z)

Branch created 

>    remember to add to charmcraftcache-hub: [canonical/data-platform-workflows@main/.github/workflows/build_charm.md](https://github.com/canonical/data-platform-workflows/blob/main/.github/workflows/build_charm.md?rgh-link-date=2024-08-08T09%3A08%3A50Z)

-> https://github.com/canonical/charmcraftcache-hub/issues/81

File: tox.ini [hunk 1]
```diff
@@ -18,19 +18,21 @@ set_env =
 allowlist_externals =
     poetry
     
-[testenv:{build,pack-wrapper}]
+[testenv:build-{production,dev,wrapper}]
 # Wrap `charmcraft pack`
 pass_env =
-    CRAFT_SHARED_CACHE
+    CI
```
Comment (carlcsaposs-canonical) [inline tox.ini:24]:
```suggestion
    CI
    GH_TOKEN
```

**Review by Mehdi-Bendriss — APPROVED** [review]

**Comment by Mehdi-Bendriss** [comment]
@Gu1nness should we have a validation of the existence `requirements.txt` in charmcraft yaml? like [here](https://github.com/canonical/opensearch-operator/blob/main/charmcraft.yaml#L19-L25)

**Review by MiaAltieri — APPROVED** [review]

**Pull request merged**